### PR TITLE
Adding backup and restore class skeleton

### PR DESCRIPTION
**Pull request by branupama**

### CODE REVIEW

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: include/backup_restore.hpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+     * This API is to get backup and restore's EEPROM or
+     * inventory path information present in the backup and restore config JSON.
+     *
+     * @param[in] i_tag - source or destination tag.
+     * @param[out] o_location - Holds hardware or inventory string value, for
+     * given i_tag.
+     * @param[out] o_path - Holds EEPROM or inventory path, for given
+     * i_tag.
+     */
+    void getBackupRestorePathInfo(const std::string& i_tag,
```
[the remark this answers is not in the document]
Comment (SunnySrivastava1984) [reply include/backup_restore.hpp:50]: Ideally, system VPD collection is IBM specific functionality and this makes the functionality specific to that. So if I don't have system VPD collection process and wants to do the collection at end with back up on cache. This logic will not let me do that.

Lets have something like this.
A variable to hold info whether backup and restore has been performed or not and a setter public method for that variable should do. 
Based on that variable this class should decide if the functionality should be performed for not.

File: include/backup_restore.hpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+     * @param[in] i_sourcePath - Source EEPROM file path or inventory path.
+     * @param[in] i_dstPath - Destination EEPROM file path or inventory path.
+     */
+    void backupAndRestoreIpzVpd(types::IPZVpdMap& io_sourceVpdMap,
+                                types::IPZVpdMap& io_dstVpdMap,
+                                const std::string& i_sourcePath,
+                                const std::string& i_dstPath);
+
+    // Parsed system config JSON
+    const nlohmann::json& m_parsedSystemJson;
```
Comment (SunnySrivastava1984) [inline include/backup_restore.hpp:88]: Why we need three class level variables?
only parsed back up and restore JSON object should be sufficient for now?

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: include/backup_restore.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (branupama) [reply include/backup_restore.hpp:50]: ok, I have move it under private.

File: include/backup_restore.hpp (hunk 2 above)
Comment (branupama) [reply include/backup_restore.hpp:88]: removed as member variable. if required in later, we can bring it back.

File: include/backup_restore.hpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    void backupAndRestoreIpzVpd(types::IPZVpdMap& io_sourceVpdMap,
+                                types::IPZVpdMap& io_dstVpdMap,
+                                const std::string& i_sourcePath,
+                                const std::string& i_dstPath);
+
+    // Parsed backup restore JSON file.
+    nlohmann::json m_parsedJson{};
+
+    // holding backup restore config JSON file path.
+    std::string m_configJsonPath;
```
Comment (SunnySrivastava1984) [inline include/backup_restore.hpp:91]: Not required.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: src/backup_restore.cpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
+#include "backup_restore.hpp"
+
+namespace vpd
+{
+BackupRestore::BackupRestore(const nlohmann::json& l_parsedSystemJson)
```
Comment (SunnySrivastava1984) [inline src/backup_restore.cpp:5]: Why it is "l_" in parameter.

File: include/backup_restore.hpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+     * @param[in] i_sourcePath - Source EEPROM file path or inventory path.
+     * @param[in] i_dstPath - Destination EEPROM file path or inventory path.
+     */
+    void backupAndRestoreIpzVpd(types::IPZVpdMap& io_sourceVpdMap,
+                                types::IPZVpdMap& io_dstVpdMap,
+                                const std::string& i_sourcePath,
+                                const std::string& i_dstPath);
+
+    // Parsed backup restore JSON file.
+    nlohmann::json m_parsedJson{};
```
Comment (SunnySrivastava1984) [inline include/backup_restore.hpp:73]: nit: to avoid any confusion lets rename this to m_backupAndRestoreParsedJson.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: src/backup_restore.cpp (hunk 4 above)
Comment (branupama) [reply src/backup_restore.cpp]: Done

File: include/backup_restore.hpp (hunk 5 above)
Comment (branupama) [reply include/backup_restore.hpp]: Done

**Review by SunnySrivastava1984 — APPROVED** [review]

File: include/backup_restore.hpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+     * Note: This API works on the keywords declared in the backup and restore
+     * config JSON. Restore or backup action could be triggered for each
+     * keyword, based on the keyvalue present in the source and destination
+     * keyword.
+     *
+     * Restore source keyword value with destination keyword value,
+     * when source keyword has default value but
+     * destination has non default value.
+     *
+     * Backup the source keyword value on destination,
```
Comment (jinuthomas) [inline include/backup_restore.hpp:50]: nit: to the destination's keywords value

File: include/backup_restore.hpp [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+     * @brief An API to backup and restore VPD.
+     *
+     * Note: This API works on the keywords declared in the backup and restore
+     * config JSON. Restore or backup action could be triggered for each
+     * keyword, based on the keyvalue present in the source and destination
+     * keyword.
+     *
+     * Restore source keyword value with destination keyword value,
+     * when source keyword has default value but
+     * destination has non default value.
```
Comment (jinuthomas) [inline include/backup_restore.hpp:48]: nit: destination's keyword

File: include/backup_restore.hpp [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+     *
+     * Backup the source keyword value on destination,
+     * when source keyword has non default value but
+     * destination has default value.
```
Comment (jinuthomas) [inline include/backup_restore.hpp:52]: nit: destination's keyword

File: include/backup_restore.hpp [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+     *
+     * @return Tuple of updated source and destination VPD map variant.
```
Comment (jinuthomas) [inline include/backup_restore.hpp:54]: So , you will always get two valid maps correct

File: include/backup_restore.hpp [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    std::tuple<types::VPDMapVariant, types::VPDMapVariant> backupAndRestore();
+
+  private:
+    /**
+     * @brief An API to handle backup and restore IPZ VPD.
```
Comment (jinuthomas) [inline include/backup_restore.hpp:60]: of IPZ type VPD.

File: include/backup_restore.hpp [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+     * @param[in,out] io_dstVpdMap - Destination VPD map.
+     * @param[in] i_sourcePath - Source EEPROM file path or inventory path.
+     * @param[in] i_dstPath - Destination EEPROM file path or inventory path.
+     */
+    void backupAndRestoreIpzVpd(types::IPZVpdMap& io_sourceVpdMap,
+                                types::IPZVpdMap& io_dstVpdMap,
+                                const std::string& i_sourcePath,
+                                const std::string& i_dstPath);
+
+    // Parsed backup restore JSON file.
```
Comment (jinuthomas) [inline include/backup_restore.hpp:72]: nit: backup and restore's JSON config file.

File: include/backup_restore.hpp [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+     * @brief An API to handle backup and restore IPZ VPD.
+     *
+     * @param[in,out] io_sourceVpdMap - Source VPD map.
+     * @param[in,out] io_dstVpdMap - Destination VPD map.
+     * @param[in] i_sourcePath - Source EEPROM file path or inventory path.
+     * @param[in] i_dstPath - Destination EEPROM file path or inventory path.
+     */
+    void backupAndRestoreIpzVpd(types::IPZVpdMap& io_sourceVpdMap,
+                                types::IPZVpdMap& io_dstVpdMap,
+                                const std::string& i_sourcePath,
```
Comment (jinuthomas) [inline include/backup_restore.hpp:69]: why i_sourcePath could be i_srcPath.  

same for io_srcVpdMap

File: include/backup_restore.hpp [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+    /**
+     * @brief An API to backup and restore VPD.
+     *
+     * Note: This API works on the keywords declared in the backup and restore
+     * config JSON. Restore or backup action could be triggered for each
+     * keyword, based on the keyvalue present in the source and destination
+     * keyword.
+     *
+     * Restore source keyword value with destination keyword value,
```
Comment (jinuthomas) [inline include/backup_restore.hpp:46]: nit: keyword's

File: include/backup_restore.hpp (hunk 11 above)
Comment (branupama) [reply include/backup_restore.hpp:72]: done

File: include/backup_restore.hpp (hunk 12 above)
Comment (branupama) [reply include/backup_restore.hpp:69]: Updated

File: include/backup_restore.hpp (hunk 10 above)
Comment (branupama) [reply include/backup_restore.hpp:60]: done

File: include/backup_restore.hpp (hunk 8 above)
Comment (branupama) [reply include/backup_restore.hpp:52]: Done

File: include/backup_restore.hpp [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+     * @brief Defaul destructor.
+     */
+    ~BackupRestore() = default;
+
+    /**
+     * @brief An API to backup and restore VPD.
+     *
+     * Note: This API works on the keywords declared in the backup and restore
+     * config JSON. Restore or backup action could be triggered for each
+     * keyword, based on the keyvalue present in the source and destination
```
[the remark this answers is not in the document]
Comment (branupama) [reply include/backup_restore.hpp:43]: Done

File: include/backup_restore.hpp (hunk 7 above)
Comment (branupama) [reply include/backup_restore.hpp:48]: Done

File: include/backup_restore.hpp (hunk 6 above)
Comment (branupama) [reply include/backup_restore.hpp:50]: Done

**Review by SunnySrivastava1984 — APPROVED** [review]

**Pull request merged**